### PR TITLE
GC: cache the largest free chunk of pages in LargeAllocPool

### PR DESCRIPTION
**Pull request by rainers**
also refactored LargeAllocPool.allocPages for better readibility and performance.

This dropped execution time of the rand_large benchmark from >3sec to 0.8 sec on Win32. It didn't have much influence on the other tests, though.

### CODE REVIEW

File: src/gc/gc.d [hunk 1]
```diff
@@ -2911,43 +2915,37 @@ struct LargeObjectPool
...
-        size_t n2;
+        if(largestFree < n)
+            return OPFAIL;
 
         //debug(PRINTF) printf("Pool::allocPages(n = %d)\n", n);
-        n2 = n;
-        for (i = searchStart; i < npages; i++)
+        size_t largest = 0;
+        while (searchStart < npages && pagetable[searchStart] != B_FREE)
+            searchStart++;
```
Comment (MartinNowak) [inline src/gc/gc.d]: Can't you do the `== B_PAGE; searchStart += bPageOffsets[searchStart];` trick here too?

**Comment by MartinNowak** [comment]
Rest looks good.

File: src/gc/gc.d (hunk 1 above)
Comment (rainers) [inline src/gc/gc.d]: Good idea, updated. I also noticed that `freePages` is a better estimate than `npages` when invalidating `largestFree`.

**Comment by MartinNowak** [comment]
Auto-merge toggled on

**Pull request merged**